### DISCUSSION TITLE
Breach Method Callback

### DISCUSSION THREAD

**Discussion by umutcakan** [yasserfarouk/scml#63] [collusion-qa] [answered]
Is there a way to understand the breached contracts where my agent is the perpetrator? During negotiation, I receive only the contracts in the on_contracts_breached method where the opponent agent is the perpetrator. If I am not mistaking all of our contracts are falling in the on_contract_executed method.

**Comment by yasserfarouk** [comment, marked as answer]
Are you sure you are committing breaches? 
You should receive on_contract_breached() even if you are the perpetrator. I will double check and return to you.

**Reply by umutcakan** [reply]
I am logging inventory, breach, and the execution of the contracts, even if I can't manage to meet the requirements, it comes to the on_contract_executed method, and when the opponent breaches, it is working as intended.

**Reply by yasserfarouk** [reply]
Because insufficient products breaches are covered by buying from the spot market except when the agent has no money to do that, agents only commit them at the time they go bankrupt. At this point, the agent cannot really do anything so it is not told that it became bankrupt as it cannot use this information.

We added `spot_market_quantity` and `spot market_loss` to the AWI to allow agents to get information about the quantities they bought from the spot market (at any point in the simulation) and the corresponding loss (i.e. increase in price when buying from the spot market again). 

We also added `is_bankrupt` to the AWI to explicitly check whether an agent (including self) is bankrupt.
